### ISSUE TITLE
System git config file is not evaluated

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#10914]
### Environment

- Git Extensions v4.0.2 / 33.33.33
- Build be087f852f799476badc2220bd86c64ad1c74c0b
- Git 2.40.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.16
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Issue description

GE asserts that the git config "core.autocrlf" was "not set" although it *is* set in the system config file "C:/Program Files/Git/etc/gitconfig".

![grafik](https://user-images.githubusercontent.com/36601201/234284646-0cf4ed67-6310-48df-8eae-e4596426d59f.png)

### Steps to reproduce

install git and GE on a new machine

### Did this work in previous version of GitExtensions?

no

### Diagnostics

```
$ git config --show-origin core.autocrlf
file:C:/Program Files/Git/etc/gitconfig true

$ cat 'C:/Program Files/Git/etc/gitconfig'
[http]
        sslCAInfo = C:/Program Files/Git/mingw64/etc/ssl/certs/ca-bundle.crt
        sslBackend = openssl
[diff "astextplain"]
        textconv = astextplain
[filter "lfs"]
        clean = git-lfs clean -- %f
        smudge = git-lfs smudge -- %f
        process = git-lfs filter-process
        required = true
[credential]
        helper = manager
[core]
        editor = \"C:\\\\Program Files\\\\Notepad++\\\\notepad++.exe\" -multiInst -notabbar -nosession -noPlugin
        autocrlf = true
        fscache = true
        symlinks = false
[pull]
        rebase = false
[credential "https://dev.azure.com"]
        useHttpPath = true
[init]
        defaultBranch = master
```

**Comment by mstv** [comment]
Should we add a (read-only) settings source "System"?

**Comment by RussKie** [comment]
Why read-only though? A user can change the system wide settings at will

**Comment by gerhardol** [comment]
System should be shown

> Why read-only though? A user can change the system wide settings at will
> 

Only as admin

**Comment by mstv** [comment]
Read-only because C:/**Program Files**/Git/etc/gitconfig would require raised privileges.

**Comment by RussKie** [comment]
That's a good point. Though, if a user isn't a local admin then we won't be
able to run a command (we'll get something like 'Access Denied', which we
(in theory) should be able to handle), which makes the settings "read
only", no?

**Comment by RussKie** [comment]
>  Read-only because C:/*Program Files*/Git/etc/gitconfig would require
raised privileges.

I never knew that config existed... O.O

**Comment by gerhardol** [comment]
worktree settings (`--worktree` if `extensions.worktreeConfig` is set) is missing too

**Comment by mstv** [comment]
> worktree settings (`--worktree` if `extensions.worktreeConfig` is set) is missing too

Yes, I have seen it. Adding such a settings page is not as straightforward as I thought. I'll need to take time for some more debugging.

> I never knew that config existed...

Me too. `git config --show-scope -l` revealed it.

**Comment by mstv** [comment]
WSL settings are ignored, too.
I think we should implement Jay's proposal, i.e. call "git config [--\<scope>]" in order to retrieve all settings, cache them and invalidate the cache on `ApplicationActivated`.
Extending `ConfigFileSettingsSet` would be error-prone. My straight-forward attempt failed.
Though `DistributedSettingsSet.DistributedSettings` ("Distributed with current repo") seems to be a GE extension, i.e. not a built-in feature of git. "git config --file \<filename>" might be used.

**Comment by gerhardol** [comment]
> WSL settings are ignored, too.

WSL git is a second class citizen still.
(I use GE on WSL 80 % of the time, but WSL works good enough for me.)
A problem with WSL is how to present the settings, many settings are out of a .git directory.
Note also that file monitoring is not working in WSL file systems.

I still believe that the WSL files and system can be monitored as current settings.
However, the related discussion to this issue demotivated me quite a lot, I have been much less active since this. Just getting close to this discussion again brings back that feeling.

**Comment by mstv** [comment]
> I use GE on WSL 80 % of the time, but WSL works good enough for me.

I know. I have just started using it - with a lot of large submodules. Once cloned, it works well enough - but only after several manual `git config` and retries.

> However, the related discussion to this issue demotivated me quite a lot, I have been much less active since this. Just getting close to this discussion again brings back that feeling.

Yes, the discussion definitely was not nice style and did not stay on the topic nor did address concerns. I had already stopped taking part for that reason before you.
But the SettingSets are a technical dept we need to pay off...

> A problem with WSL is how to present the settings, many settings are out of a .git directory.

The WSL git is run nearly always - except for the first two calls IIRC, (i.e. not for "git --version").
That's why I think using "[wsl] git config" might be a pragmatic solution.

> Note also that file monitoring is not working in WSL file systems.

Thank you for letting me know this again before I wonder why it doesn't work!

**Comment by vbjay** [comment]
> Yes, the discussion definitely was not nice style 

Imagine the huge frustrating feeling of being dismissed and trying to get the core team to see what I was seeing. So I went aggravated and pushy.

**Comment by gerhardol** [comment]
For sure, WSL Git support can be improved which I endorse - I just explained why I have not done it.

> > I use GE on WSL 80 % of the time, but WSL works good enough for me.
> 
> I know. I have just started using it - with a lot of large submodules. Once cloned, it works well enough - but only after several manual `git config` and retries.

Sidenote: I did a setup script for Git that I force everyone to use that e.g. copies ssh keys from Windows and sets aliases etc, so less difference for me...

> But the SettingSets are a technical dept we need to pay off...

Yes
Settings in WSL should be presented too. But for what instances? I have old wsl I no longer use, Docker Desktop adds some too.

> > A problem with WSL is how to present the settings, many settings are out of a .git directory.
> 
> The WSL git is run nearly always - except for the first two calls IIRC, (i.e. not for "git --version"). That's why I think using "[wsl] git config" might be a pragmatic solution.

`wsl git --version` runs when a repo in WSL is first opened.

All values in one command will only add one delay and could be done in parallel to rendering the graph to not delay commands.
For me with a repo with few settings, the command requires 200 ms (add time for parsing) compared to 100 ms for just one setting.
wsl git config will work

**Comment by astos-marcb** [comment]
In _portable_ and (if they will work at any time) _user_ installs, the system settings file **would** be changeable with _regular_ user permissions.

So GE would need to present the additional config scopes bringing this from 2 to 4 (workspace, local, global, system).

GE settings are (in theory) also possible for `workspace` and `system` scopes, raising that level count from 3 to 5.

It's likely a good idea to use the output of `git config --show-scope --show-origin -l -z` to consistently query the actual **files** used by the _configured_ Git instance.
And also to resolve the host location in WSL mode via `wsl wlspath -w <file>`.
Unclear that to do with GE settings in different scopes:
- Always follow a _standard_ directory layout (likely the better option) or
- use a file relative to the currently used Git instance config locations.

Potentially there  will be some hacks needed for libgit2-based operations if those use a different set of config files by default.
